### ISSUE TITLE
patching a common sub-dependency

### ISSUE THREAD

**Issue by abathur** [DavHau/mach-nix#234]
I've finally gotten around to giving mach-nix a second try. I got frustratingly close to something that might work for our case, but it ultimately seems to be falling over around the need to patch a common dependency.

I've boiled this down into a fairly simple case and put it in a repo with CI.

1. Here's an initial failure: [commit](https://github.com/abathur/flaskmvp/commit/1a26069d1ff9796ab27a80e6226883cf137a8b8a) | [CI](https://github.com/abathur/flaskmvp/runs/1882993899?check_suite_focus=true#step:4:1421). 
    - This just shows some dumb errata about flask/werkzeug + nix. In debug mode, werkzeug assumes it can re-run $0 as python. In our case, the flask executable has a bash wrapper so it breaks as shown here. The Nix community convinced werkzeug to adopt a fix at one point, but it was causing trouble for other users and it was reverted.
2. Since this hasn't been fixed in Nixpkgs, I've just been patching werkzeug for our dev environment. The patch is fairly simple, but doing so adds another werkzeug to the environment: [commit](https://github.com/abathur/flaskmvp/commit/a133e018f7e1627ac22c325c9a5ea5fa5d34efb9) | [CI](https://github.com/abathur/flaskmvp/runs/1883012744?check_suite_focus=true#step:4:1847)

Our actual case has fair number of packages that all depend on flask and thus on werkzeug. I may be missing a better approach, but after spending a while playing whack-a-mole with these issues it started looking like I'd end up with a translation of our pypi2nix-based requirements.nix instead of a thin layer over requirements.txt files.

**Comment by DavHau** [comment]
Hey, thanks for this very detailed report. Stuff like this helps a lot.
Multiple werkzeug ending up in the final derivation should not happen. It is a bug.
My guess is, since mach-nix is building on top of nixpgks, it needs to detect existing dependencies inside nixpkgs and replace them to prevent collisions. Sometimes this detection fails.
I'll take a look into your specific case

**Comment by abathur** [comment]
Cool. Glad to hear that, actually. I was afraid this would be a "yes, but only after some rearchitecting" kind of thing. :)

**Comment by abathur** [comment]
I'm tempted to also open another report, but I can't quite tell if this is the same issue as this, or not?

While working on this, I also found myself wishing there I had a simple idiom for specifying a `buildPythonPackage` that would use an existing resolution/context as its starting point.

I'm not sure how much sense that makes, so a little context:
- We've got a handful of requirements (common + {dev,production}) files composed together with the `-r <file>` syntax, so I used `cat` to append them and grep to filter out those statements. A lot of these requirements (including Flask) are pinned to specific versions as small breaks have accumulated over several years in production.
- We've got a private package on gemfury that is also referenced from the requirements. I separated it out into its own `mach-nix.buildPythonPackage` invocation to provide a source, and add it to `overridesPre`. It also depends on Flask, and a few other packages, but none of these specify a version.
- When we build this, the private package itself gets resolved in correctly, but I end up with two Flask packages, one of the latest version. I assume mach-nix tries to resolve this private package's requirements from scratch and in isolation, when pip would normally resolve them in the broader context, respecting the pinned versions.

**Comment by DavHau** [comment]
Sorry for the delay. I improved the mechanism of overriding nixpkgs via 6473903da7e17b11f7c41f5729713bf46f84e6c2.
The general problem is, that when overriding python modules from nixpkgs, they have python dependencies from nixpkgs, which need to be replaced with the python packages defined by mach-nix. The previous replacement function didn't replace sub-dependencies of these dependencies. Now I updated it to work recursively through all dependencies.
Your expression now works if you use the master branch of mach-nix.

> When we build this, the private package itself gets resolved in correctly, but I end up with two Flask packages, one of the latest version (and one of the pinned version). I assume mach-nix is resolving this private package's requirements from scratch and in isolation, when pip would normally resolve them in the broader context, respecting the pinned versions?

All packages should always be resolved together, even if some of them are included via packagesExtra. If that is not the case, it is a bug. Maybe you were just running into the same issue which just got resolved. If not, I kindly ask you to open another issue with the expression triggering the collision. Thanks

**Comment by abathur** [comment]
Sorry about the delay on my end as well; had a rough week of Texas weather.

It looks like this works. I've tried the simple case:
- I updated to master in https://github.com/abathur/flaskmvp/commit/cc7d85df8621992003cedd7867ec77bd6ec08f3b
- the pythonCatchConflictsPhase ran clean https://github.com/abathur/flaskmvp/runs/1962757835?check_suite_focus=true#step:4:1605 (whole run failed because it runs the server and I had to cancel it)

I'll try to revisit the complex version I was working on today.
